src/libexpr/eval-cache.cc

```diff
@@ -764,4 +764,32 @@ StorePath AttrCursor::forceDerivation()
     return drvPath;
 }
 
+std::string showAttrValueType(const AttrValue & attrValue)
+{
+    return std::visit(overloaded{
+        [&](const string_t & s) { return "a string"; },
+        [&](bool) { return "a boolean"; },
+        [&](int_t) { return "an integer"; },
+        [&](const std::vector<std::string> &) { return "a list of strings"; },
+        [&](const std::vector<Symbol> &) { return "an attribute set"; },
+        [&](const placeholder_t &) { return "a placeholder"; },
+        [&](const missing_t &) { return "nothing"; },
+        [&](const misc_t &) { return "an unknown value"; },
+        [&](const failed_t &) { return "an error"; },
```

Comment:
@roberth would the below be better you think?

```suggestion
        [&](const missing_t &) { return "nothing (try with `--option eval-cache-false` for more information)"; },
        [&](const misc_t &) { return "an unknown value (try with `--option eval-cache-false` for more information)"; },
        [&](const failed_t &) { return "an error (try with `--option eval-cache-false` for more information)"; },
```

It's a bit hacky since that will get interpolated in the final error message, but the final result should be OK:

`error: expected a function, but got an unknown value (try with `--option eval-cache-false` for more information)`

Reply:
That should be `--no-eval-cache`, there is no `eval-cache-false`.

But a better fix is to make sure to use `maybeGetAttr(..., true)` to force evaluation of failed attributes.